사용자는 식품 상세 정보를 조회한다

this-is-spear opened the pull request:
## 작업에 대한 간단한 요약

사용자는 식품 상세 정보를 조회할 수 있습니다.  Food는 공공데이터에서 제공하는 출력 데이터를 기준으로 구현했습니다.

<img width="942" alt="스크린샷 2023-07-14 오전 3 55 33" src="https://github.com/towards-sweetly/service-be/assets/92219795/c5bafacf-8d4e-48f8-b753-e002a667fda0">

자세한 내용은 [이슈](#3)에서 확인할 수 있습니다.

<br><br>

### 변경내역

- Food 구현
- FoodQueryService 구현
- FoodController 구현

<br><br>

### 어떤부분에 집중해서 리뷰하면좋을지

#### Food

##### 1️⃣ 상태

Food 엔티티에서 필요한 상태는 오버플로우 및 언더플로우를 염두해 BigDecimal로 구현했습니다. 하지만 단위가 전부 다른 문제는 Gram, MilliGram, Kcal를 이름 뒷편에 추가했지만 가독성이 떨어지는 문제가 발생했습니다.

<img width="927" alt="스크린샷 2023-07-14 오전 3 48 52" src="https://github.com/towards-sweetly/service-be/assets/92219795/5176c0dd-1b9b-47a6-b524-63408e59652c">

<br><br>

추가로 BigDecimal로 구현한 상태가 너무 많아 값을 잘못 입력할 경우가 발생할 수 있습니다. 이에 대한 대비책으로 상태를 한 계층 Wrapping 해볼 염두를 하고 있습니다.

<img width="956" alt="스크린샷 2023-07-14 오전 3 50 03" src="https://github.com/towards-sweetly/service-be/assets/92219795/4075f778-ffc4-4696-ab4c-8a1f363b8e0d">

<br>

##### 2️⃣ 생성자

생성자 관리 방법은 모든 상태를 입력하는 생성자를 통해 인스턴스가 생성되도록 구현했습니다. 새로운 엔티티를 저장하는 일이 없어 효과가 없지만 구현 비용 없이 수정 비용을 줄일 수 있어서 채택했습니다.

<img width="958" alt="스크린샷 2023-07-14 오전 3 59 45" src="https://github.com/towards-sweetly/service-be/assets/92219795/0990dccc-7e21-4be0-9d15-083ce917011f">

<br>

##### 3️⃣ 테스트

생성자 및 게터를 테스트했습니다. 구현에서 동일한 유형의 파라미터로 인해 잘못된 주입이 발생할 경우를 대비했습니다.

<img width="912" alt="스크린샷 2023-07-14 오전 4 02 19" src="https://github.com/towards-sweetly/service-be/assets/92219795/0a653cb6-26f3-4d90-b403-0be29b3dc47a">

<br><br>

### 이슈번호 첨부,  기타내용

- 저장소 구현은 진행하지 않았습니다.

this-is-spear committed:
feature : Food 구현 및 테스트

###Description###

동일한 유형의 파라미터로 인해 잘못된 주입이 발생할 경우를 대비해 생성자 및 게터 Test 진행

this-is-spear committed:
feature : FoodQueryService 구현 및 테스트

###Description###

동일한 유형의 파라미터로 인해 잘못된 주입이 발생할 경우를 대비해 생성자 및 게터 Test 진행

this-is-spear committed:
feature : FoodController 구현 및 테스트

###Description###

- Controller 테스트 구현 및 진행
- 생성한 FoodDetailRequest 제거
- Food 테스트 패키지에 Stub 추가
- FoodDetailResponse 매핑 추가

youngreal commented on src/test/java/com/management/servicebe/food/ui/FoodControllerTest.java:
```diff
@@ -0,0 +1,39 @@
+import static org.mockito.Mockito.when;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+
+@ExtendWith(MockitoExtension.class)
```
해당 테스트클래스에서는 @MockBean만 사용하게 될것같은데 그렇다면 해당코드는 필요없어 보입니다
혹시 달아주신 다른이유가 있으실까요???

this-is-spear replied:
실수로 추가됐습니다. 수정하겠습니다 ☺️

youngreal commented on src/test/java/com/management/servicebe/food/ui/FoodControllerTest.java:
```diff
@@ -0,0 +1,39 @@
+
+    @Test
+    void findFoodDetail() throws Exception {
+        when(foodQueryService.findFoodDetail(1L)).thenReturn(FoodDetailResponse.of(바나나칩));
```
해당 테스트에서의 when은 mockMvc.perform 부분이라고 생각되서 
when()도 좋지만 given()을 이용하면 좀더 g/w/t 양식으로 더 가독성있게 읽힐것 같은데 어떠신가요??

this-is-spear replied:
`given-when-then`에 맞게 진행하면 코드를 서로 이해하기 쉽겠군요.
`BDDMockito` 라이브러리를 사용해서 진행하겠습니다 ☺️

youngreal commented on src/test/java/com/management/servicebe/food/application/FoodQueryServiceTest.java:
```diff
@@ -50,7 +33,7 @@ void findFoodDetail() {
         when(foodRepository.findById(id)).thenReturn(Optional.of(바나나칩));
```
해당 부분도 when대신 given사용하면 좋을것같습니다 ㅎㅎ

youngreal commented on src/main/java/com/management/servicebe/food/application/FoodQueryService.java:
```diff
@@ -12,11 +11,8 @@ public class FoodQueryService {
-                food.getSaturatedFattyAcidsGram(), food.getTransFattyAcidsGram());
+    public FoodDetailResponse findFoodDetail(final Long id) {
+        final var food = foodRepository.findById(id).orElseThrow(IllegalArgumentException::new);
+        return FoodDetailResponse.of(food);
```
저는 request/response dto는 HTTP환경에 필요한 의존적인 dto라고 생각해서
 
비즈니스로직에서 HTTP환경에 필요한 응답 dto로 변경해주는것보다는 

컨트롤러에서 이 일을 해주는게 조금더 적합할수있지않을까? 싶은 생각이드는데 요 부분은 어떻게 생각하시나요?? 

상위 dto패키지에(app/dto) response를 두는것도 좋지만 ui/response 같은패키지를 만들어서 입출력 역할을 담당하는 ui쪽으로 옮기는것은 어떠실까요??

this-is-spear replied:
클라이언트에 맞게 원하는 데이터를 추출하는 역할을 비즈니스 역할이라 생각했고, 도메인 엔티티가 프레젠테이션 레이어에 노출이 안됐으면 좋겠다는 생각을 하고 있었습니다.

HTTP 환경에 필요한 정보보다 클라이언트가 필요한 정보로 보는게 맞다는 생각이 듭니다. 또한 프레젠테이션, 비즈니스, 도메인, 인프라 레이어는 한 계층만 의존할 수 있도록 의존 방향을 단순화하는 건 어떨까요?

youngreal replied:
> 클라이언트에 맞게 원하는 데이터를 추출하는 역할을 비즈니스 역할이라 생각했고

오 저랑 생각이 다르시군요 저는 요청에대한 응답을 가공하는 과정은 스프링MVC에 특화된 컨트롤러의 역할이라고 생각합니다
Response라는건 HTTP환경에서만 필요한 표현이 아닌가 싶기도합니다 

> 도메인 엔티티가 프레젠테이션 레이어에 노출이 안됐으면 좋겠다는 생각을 하고 있었습니다.

저는 오히려 도메인을 최대한 활용해야한다고 생각해서 개인적으로 프레젠테이션 레이어에 노출이 안됐으면 하는 이유가 궁금합니다

>또한 프레젠테이션, 비즈니스, 도메인, 인프라 레이어는 한 계층만 의존할 수 있도록 의존 방향을 단순화하는 건 어떨까요?

해당부분은 조금 이해를 못했는데 예를들어 app/dto 같이 공통 dto패키지에 있는 Response에만 의존할수있도록 단순화한다는 말씀이 맞으실까요?

현재 구조에서는 응답이라는 response가 수정되면 controller/service 둘다 수정이 발생하는데 저는 응답이라는것 자체가 서비스로직이라고 생각하진않았던것같습니다
(이부분은 건창님이 클라이언트에 맞게 원하는 데이터를 추출하는 것도  비즈니스역할이라고 보셨으면 그럴수있다고 생각합니다ㅎㅎ)

해당부분에 대해 조금더 설명해주시면 좋을것같습니다!


또 하나 염려되는점은 만약 FoodQueryService를 다른 Service에서 사용한다면 그 Service또한 Response라는 DTO에 묶이게되는데 

저는 이런걸고려했을때 Service에서는 최대한 순수한 도메인엔티티를 넘겨주고 필요한 곳에서 DTO로 변환하는게 좋을수있다고 생각합니다.

youngreal commented on src/main/java/com/management/servicebe/food/application/FoodQueryService.java:
```diff
@@ -12,11 +11,8 @@ public class FoodQueryService {
-                food.getCaloriesKcal(), food.getCarbohydratesGram(), food.getProteinGram(), food.getFatGram(),
-                food.getSugarsGram(), food.getSodiumMilliGram(), food.getCholesterolMilliGram(),
-                food.getSaturatedFattyAcidsGram(), food.getTransFattyAcidsGram());
+    public FoodDetailResponse findFoodDetail(final Long id) {
```
메서드 파라미터나 객체에 final을 붙이는건 컨벤션의 영역일것같은데 붙이는걸 컨벤션으로 알고있으면 될까요?

this-is-spear replied:
좋습니다. 두 명이 작성하는 거니 조금 더 방어적으로 작성해나가죠 ☺️

youngreal commented on src/test/java/com/management/servicebe/food/domain/FoodTest.java:
```diff
@@ -0,0 +1,61 @@
+
+class FoodTest {
+
+    @Test
+    @DisplayName("""
+            식품은 식품 이름, 카테고리, 1회 제공량(g), 열량(kcal), 탄수화물(g), 단백질(g), 지방(g), 당류(g), 나트륨(mg), 콜레스테롤(mg), 포화지방산(g), 트랜스지방산(g)으로 구성된다.
+            """)
```
> 생성자 및 게터를 테스트했습니다. 구현에서 동일한 유형의 파라미터로 인해 잘못된 주입이 발생할 경우를 대비했습니다.

오 해당 테스트로 만약 탄수화물에 지방g이 들어가면 테스트가 실패하게되어 막을수있다는 걸까요?

this-is-spear replied:
아닙니다 ☺️ 제가 생성과정에서 잘못된 매핑 과정으로 구현할까봐 걱정되서 테스트를 작성했습니다. 

말씀하신 부분이 실현되려면 각 상태를 한 계층 래핑할 필요가 있어보입니다. 래핑해서 상태 관리를 할 필요가 있을까요? 어떻게 생각하시나요

youngreal replied:
저는 딱히 필요없다고 생각합니다 잘못주입 방지를 위해 생성자보다 빌더를 써서 실수를 줄일수 있을것 같기도하고 코드리뷰도 있으니 악의적으로 순서를 바꾸는거 아닌이상 잡을수 있는 버그로 보입니다

youngreal reviewed (commented):
PR한번에 동일파일에서 변경내역이 생겨서 FoodQueryService는 모든 커밋을 보고 리뷰를 달아야해서 커밋이 좀더 많아지면  단점 커질수 있을것같습니다.

커밋메시지 본문도 작성해주셨는데 https://cbea.ms/git-commit/  해당글 7번에 해당하는 내용 한번 읽어보시면 좋을것같습니다 ㅎㅎ

저도 글을 참고하고 적용하고 코드리뷰시에 매우 용이했던것 같습니다.

고생하셨습니다!!

this-is-spear committed:
refact : FoodControllerTest 수정

###Description###

- 사용하지 않는 익스텐션 삭제
- Mockito -> BDDMockito 라이브러리로 변경
- var 타입 활용

this-is-spear committed:
refact : FoodQueryServiceTest 수정

###Description###

- 가독성을 위해 Mockito.when() -> BDDMockito.given() 으로 변경

this-is-spear committed:
feature : InMemoryFoodRepository 구현

###Description###

- InMemoryFoodRepository 구현
- 더미 데이터 추가를 위한 StorageBootstrap 구현

this-is-spear committed:
chore : build.gradle 수정

###Description###

- 사용하지 않는 의존성 라이브러리 삭제

this-is-spear commented:
@youngreal 말씀하산 수정이 필요한 부분은 변경했고 추가적인 작업도 진행했습니다.

그리고 이야기 했던 `더미 데이터를 넣어 간단하게 테스트할 수 있는 환경을 구성하자`라는 내용을 기반으로 코드를 구현했습니다. 한 번 확인 부탁드릴게요!

youngreal commented:
확인했습니다! 저였으면 나중에 또 손한번가니까 바로 DB에 때려박는방식으로 했을텐데 꼼꼼하시네요 ㄷㄷ

Commit:
refact : fatGram 검증 과정 추가

Commit:
refact : StorageBootstrap 수정

###Description###

프로토타입 실행을 위한 더미데이터 추가
